File: 513-find-bottom-left-tree-value/ken/513.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def findBottomLeftValue(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        cur = [root]
        left = root.val

        while len(cur):
            next = []
            left = cur[0].val
            for node in cur:
                if node.left:
                    next.append(node.left)
                if node.right:
                    next.append(node.right)

            cur = next

        return left

File: 513-find-bottom-left-tree-value/leochang/513.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
	int findBottomLeftValue(TreeNode* root) {
		int maxDepth = 0;
		int ans = root->val;
		findBottomLeftValue(root, 0, maxDepth, ans);
		return ans;
	}
private:
	void findBottomLeftValue(TreeNode* root, int depth, int &maxDepth, int &ans) {
		if (root == NULL) return;

		findBottomLeftValue(root->left, depth + 1, maxDepth, ans);
		findBottomLeftValue(root->right, depth + 1, maxDepth, ans);

		if (depth > maxDepth) {
			maxDepth = depth;
			ans = root->val;
		}
	}
};
